prereise/gather/hydrodata/eia/fetch_historical.py

```diff
@@ -0,0 +1,129 @@
+import re
+
+import pandas as pd
+import requests
+from tqdm import tqdm
+
+from prereise.gather.impute import linear
+
+category_template = (
+    "https://api.eia.gov/category/?api_key={api_key}&category_id={category_id}"
+)
+series_template = "https://api.eia.gov/series/?api_key={api_key}&series_id={series_id}"
+top_level_category_id = 3390101
+
+
+def parse_input_regions(available_regions, requested_regions):
+    """Parse requested regions, allowing either exact string matches or balancing
+    authority abbreviations.
+
+    :param iterable available_regions: set of available regions (str).
+    :param iterable requested_regions: set of request regions (str).
+    :return: (*set*) -- parsed regions.
+    :raises ValueError: if any requested region isn't present in the set available from
+        EIA, or matches multiple available regions.
+    """
+    parsed_regions = set()
+    for r in requested_regions:
+        if r in available_regions:
+            parsed_regions.add(r)
+            continue
+        abbreviation_matches = {a for a in available_regions if f"({r})" in a}
+        if len(abbreviation_matches) == 0:
+            print(f"No hydro data available for region {r}, skipping")
+            continue
+        if len(abbreviation_matches) > 1:
+            raise ValueError(f"Multiple regions found for {r}: {abbreviation_matches}")
+        (single_match,) = abbreviation_matches  # unpack singleton set
+        parsed_regions.add(single_match)
+    return parsed_regions
+
+
+def get_generation(api_key, year, regions=None, acceptable_missing=48, verbose=False):
+    """Fetch hourly generation by balancing authority from EIA, filter to regions with
+    a full year of hydro data for the given year.
+
+    :param str api_key: EIA api key from <https://www.eia.gov/opendata/register.php>.
+    :param int/str year: year of data to return. Any regions with too many missing
+        entries in this year will be dropped (see ``acceptable_missing``).
+    :param iterable regions: subset of regions to fetch & return. If None, all will be
+        returned.
+    :param int acceptable_missing: allowable number of missing entries, will be linearly
+        imputed. If more than this number are missing in the given ``year``, the region
+        will be dropped.
```

Comment:
I believe we are talking about number of missing hours. Correct?

Reply:
Yes. Do you think there's a way to rephrase this to be clearer?

Reply:
allowable number of missing hours (rows). Values will be linearly imputed or region will be dropped depending on number data fetched that are missing. But you are the native speaker

Reply:
Done (here and below)

Reply:
Thanks